JSDK 2337: media monitor

makarandp0 opened the pull request:
This change adds Ice media monitor - to detect the disconnects, and start reconnect early. This is mostly based on corresponding CPP work in https://github.com/twilio/rtc-cpp/pull/640

Change:
1) added a new class IceMediaMonitor `lib/signaling/v2/icemediamonitor.js` that queries the peerConnection for media stats, and calls back when it detects inactivity. (inactivity is defined as no bytes received in 3 seconds)
2) PeerConnection (lib/signaling/v2/peerconnection.js) starts media monitor when ice state is `connected`, and stops it when ice state is `closed` or `failed`. 

Checklist before merge
- [x] ~merge https://github.com/twilio/twilio-webrtc.js/pull/102, and update this change to use `timestamp`~ n/a
- [x] remove debug console.log calls, added in the change 
- [x] resolve and incorporate questions identified in https://paper.dropbox.com/doc/Ice-Media-Monitor-for-JS--AdD_RBac6xRW8YBvKcLizZ68AQ-z2sUflMxe4dXBrjfyWl42
- [x] add more unit tests to verify Monitor callbacks
- [ ] test for network change on chrome/firefox and safari

makarandp0 commented on lib/signaling/v2/icemediamonitor.js:
```diff
@@ -0,0 +1,98 @@
+   * Start monitoring.
+   * @returns {void}
+   */
+  start(onInactiveCallback) {
```
I initially had  this class as EventEmitter (instead of taking callback param here), but decided against that, as callback works better in this case because 1) we have just one listener 2) the lifetime of callback is between start and stop, and it's easier for caller to call start/stop instead of an additional on/off for event listener.

manjeshbhargav replied:
@makarandp0 I think you can convert this to a Promise-based method and get rid of the `stop()` method 
 and the `_timer` property:
```js
class IceConnectionMonitor {
  ...
  /**
   * Check if the ICE connection is inactive.
   * @returns {Promise<void>}
   */
  checkIfInactive() {
    return wait(this._activityCheckPeriodMs).then(() => {
      return this._getIceConnectionStats();
    }).then(stats => {
      if (!stats) {
        return this.checkIfInactive();
      }
      if (!this._lastActiveStats || this._lastActiveStats.bytesReceived !== stats.bytesReceived) {
        this._lastActiveStats = stats;
      }
      if (stats.timestamp - this._lastActivity.timestamp < this._inactivityThresholdMs) {
        return this.checkIfInactive();
      }
    });
  }
  ...
}

// This can be moved to lib/utils/index.js

/**
 * Wait for a given period of time.
 * @param {number} delayMs
 * @returns {Promise<void>}
 */
function wait(delayMs) {
  return new Promise(resolve => setTimeout(resolve, delayMs));
}
```

makarandp0 replied:
@manjeshbhargav, I think your suggestion would work if we decided to only start the monitor when Ice goes into disconnected state, and simply wait for 3 seconds of non-activity. But since we are monitoring at connected state, we need this callback to fire every time it detects inactivity, and peerConnection object acts on it only when it finds itself in disconnected state when callback fires. Maybe I can setup a quick meeting to syncup on this.

makarandp0 replied:
never mind my comment above - I think I better understand your proposal now, and how it would work, but I do have few concerns with it, as I commented earlier.

makarandp0 replied:
updated this now that we've decided to listen only after disconnect.

makarandp0 commented on test/unit/spec/signaling/v2/icemediamonitor.js:
```diff
@@ -0,0 +1,94 @@
+    });
+  });
+
+  describe('Callback', () => {
```
WIP

manjeshbhargav commented on lib/signaling/v2/icemediamonitor.js:
```diff
@@ -0,0 +1,96 @@
+      },
+      _inactivityThresholdMS: {
+        value: options.inactivityThresholdMS
+      }
```
@makarandp0 Can you define `_lastActiveStats` here?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -185,6 +187,13 @@ class PeerConnectionV2 extends StateMachine {
       _rtpSenders: {
         value: new Map()
       },
+      _mediaMonitorTimer: {
```
@makarandp0 Looks like this is not used anywhere.

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -99,6 +100,7 @@ class PeerConnectionV2 extends StateMachine {
       setBitrateParameters,
       setCodecPreferences,
       setSimulcast,
+      IceMediaMonitor,
```
@makarandp0 `IceConnectionMonitor`?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -7,6 +7,7 @@ const DefaultMediaStream = WebRTC.MediaStream;
 const DefaultRTCIceCandidate = WebRTC.RTCIceCandidate;
 const DefaultRTCPeerConnection = WebRTC.RTCPeerConnection;
 const DefaultRTCSessionDescription = WebRTC.RTCSessionDescription;
+const IceMediaMonitor = require('./icemediamonitor.js');
```
@makarandp0 `iceconnectionmonitor.js`?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -185,6 +187,13 @@ class PeerConnectionV2 extends StateMachine {
+        writable: true,
+        value: null
+      },
+      _mediaMonitor: {
```
@makarandp0 Can you change this to:
```js
_iceConnectionMonitor: {
  value: new options.IceConnectionMonitor(peerConnection)
}
```
?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -474,10 +483,15 @@ class PeerConnectionV2 extends StateMachine {
+    this._log.debug(`makarand: ICE connection state is "${this._lastIceConnectionState}" => "${iceConnectionState}"`);
+    // makarand: TODO: is it okay to short-circuit here? or anybody depends on
+    // iceConnectionStateChanged firing on every non-change
+    if (this._lastIceConnectionState === iceConnectionState) {
```
@makarandp0 Let's retain how it was before, unless you can demonstrate that "iceConnectionStateChanged" is emitted even when there is a non-change.

makarandp0 replied:
Looks like the code previously assumed that event could be emitted for non-changes. Notice that https://github.com/twilio/twilio-video.js/blob/master/lib/signaling/v2/peerconnection.js#L494  
explicitely ensures that old state is not "failed" 
```
    // Case 1: Transition to "failed".
    if (this._lastIceConnectionState !== 'failed' && iceConnectionState === 'failed' && !this._shouldRestartIce && !this._isRestartingIce) {
```

This change allows to get rid of that check.

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -488,10 +502,46 @@ class PeerConnectionV2 extends StateMachine {
     }
 
+    if (iceConnectionState === 'connected') {
+      this._mediaMonitor.start(() => {
```
@makarandp0 This can be:
```js
this._iceConnectionMonitor.checkIfInactive().then(() => {
  if (this._lastIceConnectionState !== 'disconnected') {
    return this._iceConnectionMonitor.checkIfInactive();
  }
  if (!this._shouldRestartIce && !this._isRestartingIce) {
    this._shouldRestartIce = true;
    this.offer();
  }
});
```

makarandp0 replied:
ah, Now I understand what you mean by using promise for this api -  Its using nested promises -  I think there are couple cons with this approach.

1. nested promises are making code bit complicated.
2. this usage makes it difficult to use "previous" state from the media monitor. For example handling of this case is not obvious.
a) connected: time 0 -  bytes = 0 - Api was called and it returned a promise..
b) connected: time 4 -  bytes = 0 - promise resolved. 
c) Api was was called again   because state is still connected.
d) disconnected:
e) disconnected: time 5 - bytes = 0.

At e) should the API resolve the promise? For callback it's obvious that it should fire again (as it was started, and callbacks on every inactivity). But since a new promise was started c) and it hasn't yet been 3 seconds of inactivity (although total inactivity is 5 seconds).


3. Also  with this promise API the promise never gets cleared (as there is no `stop` functionality) That means we will always have an outstanding promise (and associated timer) for peer connection - long after the peer connection is closed. This would lead to memory leaks, if there were lots of peerconnection on the app.

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -488,10 +502,46 @@ class PeerConnectionV2 extends StateMachine {
+          }
+        }
+      });
+    } else if (iceConnectionState === 'failed' || iceConnectionState === 'closed') {
```
@makarandp0 I don't think this will be necessary with the new way of checking inactivity, since we are using `setTimeout` instead of `setInterval`.

makarandp0 replied:
updated this now that we've decided to listen only after disconnect.

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -488,10 +502,46 @@ class PeerConnectionV2 extends StateMachine {
   }
 
+  // TODO: simplify and move to IceMediaMonitor once timestamp is added to the webrtc getStat api.
+  getMediaStats() {
```
@makarandp0 This method won't be necessary due to availability of `IceConnectionMonitor#_getIceConnectionStats()`.

manjeshbhargav commented on lib/util/constants.js:
```diff
@@ -71,3 +71,6 @@ module.exports.typeErrors = {
 module.exports.DEFAULT_NQ_LEVEL_REMOTE = 0;
 module.exports.MAX_NQ_LEVEL = 3;
+
+module.exports.ICE_MONITOR_ACTIVITY_CHECK_PERIOD_MS = 1000;
```
@makarandp0 Nitpick: I think we can shorten this to `ICE_ACTIVITY_CHECK_PERIOD_MS`.

manjeshbhargav commented on lib/util/constants.js:
```diff
@@ -71,3 +71,6 @@ module.exports.typeErrors = {
 module.exports.MAX_NQ_LEVEL = 3;
+
+module.exports.ICE_MONITOR_ACTIVITY_CHECK_PERIOD_MS = 1000;
+module.exports.ICE_MONITOR_INACTIVITY_THRESHOLD_MS = 3000;
```
@makarandp0 Nitpick: I think we can shorten this to `ICE_INACTIVITY_THRESHOLD_MS`.

manjeshbhargav commented on test/unit/index.js:
```diff
@@ -49,6 +49,7 @@ require('./spec/signaling/v2/localparticipant');
 require('./spec/signaling/v2/networkqualitymonitor');
 require('./spec/signaling/v2/networkqualitysignaling');
 require('./spec/signaling/v2/recording');
+require('./spec/signaling/v2/icemediamonitor');
```
@makarandp0 `iceconnectionmonitor.js`.

makarandp0 commented on lib/signaling/v2/icemediamonitor.js:
```diff
@@ -0,0 +1,96 @@
+
+const { ICE_MONITOR_ACTIVITY_CHECK_PERIOD_MS, ICE_MONITOR_INACTIVITY_THRESHOLD_MS } = require('../../util/constants');
+
+/**
```
okay, will add line describing the class
 ```
  Monitors a {@link RTCPeerConnection}'s ICE media stats and notifies
 caller when inactivity is detected.
```
not mentioning the ICE connection state, as that logic is in PeerConnectionV2

makarandp0 commented on lib/signaling/v2/icemediamonitor.js:
```diff
@@ -0,0 +1,96 @@
+  /**
+   * Construct a {@link IceMediaMonitor}.
+   */
+  constructor(pc, options) {
```
yes, that makes sense. I will update to use RTCPeerConnection, and also move the getMediaStats over to this file.

makarandp0 commented on lib/signaling/v2/icemediamonitor.js:
```diff
@@ -0,0 +1,96 @@
+    console.log.apply(console.log, newArgs.concat(args));
+  }
+
+  _getMediaStats() {
```
Although it may not be necessary in this case - I thought using twilio-webrtc was preferred because it ensures standardized response across browsers. But I am open to using peerConnection directly.

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -488,6 +493,19 @@ class PeerConnectionV2 extends StateMachine {
     }
 
+    if (iceConnectionState === 'disconnected') {
+      this._iceConnectionMonitor.start(() => {
```
@makarandp0 We need to make sure that when we access `PeerConnectionV2#iceConnectionState`, it is set to 'failed' here, even though the actual state is disconnected. We can do something like this:
```js
this._iceConnectionMonitor.start(() => {
  this._isIceConnectionInactive = true;
  ...
});
```
Then, we can modify the `iceConnectionState` getter for PeerConnectionV2 like this:
```js
get iceConnectionState() {
  return this._peerConnection.iceConnectionState === 'disconnected' && this._isIceConnectionInactive
    ? 'failed'
    : this._peerConnection.iceConnectionState;
}
```

makarandp0 replied:
ah, I missed this comment. makes sense.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+const IceConnectionMonitor = require('../../../../../lib/signaling/v2/iceconnectionmonitor');
+
+describe('IceConnectionMonitor', () => {
+
```
@makarandp0 Nitpick: Remove extra line?

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+
+describe('IceConnectionMonitor', () => {
+
+  describe('.constructor', () => {
```
@makarandp0 Nitpick: Remove `.`?

manjeshbhargav replied:
@makarandp0 I don't think we need to test for internal variables/methods unless absolutely required. Ideally we should just test public facing variables/methods like `start()` and `stop()`.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+    });
+  });
+
+  describe('.start', () => {
```
@makarandp0 Nitpick: `#start`?

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+  describe('.start', () => {
+    // eslint-disable-next-line no-undefined
+    ['foo', 45, { foo: 5 }, null].forEach((callback) => {
+      it(`throws if callback is ${typeof callback} `, () => {
```
@makarandp0 Not needed.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+    });
+  });
+
+  describe('._getMediaStats', () => {
```
@makarandp0 Not necessary.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+      });
+    });
+
+    it('starts the timer', () => {
```
@makarandp0 I think we can test the callback firing here, which is the only public facing behavior. We should not test internal implementation details like starting/stopping the timer.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+    });
+  });
+
+  describe('.stop', () => {
```
@makarandp0 Nitpick: `#stop`?

makarandp0 replied:
q: where do we use '.foo' v/s '#foo' ? didn't see any other usage of '#foo' in the code

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+  });
+
+  describe('.stop', () => {
+    it('stops the timer, and resets the state', () => {
```
@makarandp0 Should not be testing implementation details, only expected behavior.

makarandp0 replied:
it's a uni test, and I think it is useful to verify the internal state from unit test. But I agree overall, its better to test public interfaces. Here i did not want to add to the public interface to ensure that the timer is turned off.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+    });
+  });
+
+  describe('on inactivity', () => {
```
@makarandp0 This can move to the `#start` test.

manjeshbhargav commented on test/unit/spec/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,172 @@
+  });
+});
+
+function mockMediaStats(iceConnectionMonitor, mediaStatResults) {
```
@makarandp0 `mockIceConnectionStats` and `statsResults`?

manjeshbhargav commented on test/unit/spec/signaling/v2/peerconnection.js:
```diff
@@ -5,12 +5,34 @@ const EventEmitter = require('events');
+    IceConnectionMonitor.prototype.start.restore();
+    IceConnectionMonitor.prototype.stop.restore();
+  });
+
```
@makarandp0 Can you also add or update the test for `PeerConnectionV2#iceConnectionState` to return "failed" if the actual state is "disconnected" and ICE is inactive, and should return the underlying RTCPeerConnection's `iceConnectionState` otherwise?

makarandp0 replied:
oh, I did not think about this - Do we need to keep track of the ice failure predicted by monitor and pretend that it was real failure to the callers of PeerConnectionV2 ? Let me sync up with you  understand this better.

makarandp0 commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,130 @@
+   * @returns {void}
+   */
+  start(onInactiveCallback) {
+    if (typeof onInactiveCallback !== 'function') throw new Error('invalid callback');
```
@manjeshbhargav, I am big fan of asserts - but did not see usage in our codebase. this looks like something that we could assert for (and not have an elaborate throw like I have here). Do we use asserts in our code? basically some code that is mostly for for helping debug, catch bugs early ?

makarandp0 replied:
I will remove this check, but wanted to discuss this with you more about.

makarandp0 commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,130 @@
+   * returns media stats or null.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
+   */
+  _getMediaStats() {
```
Re: introduction of _getMediaStats - I added it to  help isolate the peerConnection.getStats logic, and thus help write effective unit tests. like "getMediaStats works fine when _getIceConnectionStats throws or returns undefined"

makarandp0 replied:
removed the method now.

makarandp0 commented:
Thank you for the feedback and catching unit test failure issue @manjeshbhargav - I think my tests were leaking a timer which was causing failures during other test runs ). I have fixed the issue, and added some instrumentation to see if it happens again. i will remove the instrumentation before merge along with other debug code.

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,130 @@
+const { ICE_ACTIVITY_CHECK_PERIOD_MS, ICE_INACTIVITY_THRESHOLD_MS } = require('../../util/constants');
+
+/**
+ * Monitors a {@link RTCPeerConnection}'s media stats and notifies
```
@makarandp0 "media" is still present.

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,128 @@
+   * returns ICE connection stats or null.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
+   */
+  _getStats() {
```
@makarandp0 I still think we don't need this method. Why would we want to copy the stats values into a JSON instead of just using the RTCIceCandidateStats directly? You can also attach the `catch()` handler to `_getIceConnectionStats()`.

makarandp0 replied:
okay, I guess the method has served its purpose of being able to test. I will removed it and fold into `_getIceConnectionStats`

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  /**
+   * Start monitoring.
```
@makarandp0 Can you add a more descriptive statement here along with specifying the parameter?
```js
/**
 * Start monitoring the ICE connection.
 * @param {function} onIceConnectionInactive
 * ...
 */
```

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,128 @@
+  }
+
+  /**
+   * Stop monitoring.
```
@makarandp0 Nitpick: `Stop monitoring the ICE connection state.`?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -290,7 +299,7 @@ class PeerConnectionV2 extends StateMachine {
    */
   get iceConnectionState() {
-    return this._peerConnection.iceConnectionState;
+    return (this._isIceConnectionInactive && this._peerConnection.iceConnectionState === 'disconnected') ? 'failed' : this._peerConnection.iceConnectionState;
```
@makarandp0 Nitpick: can you make this multiline so that its not too long?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -474,7 +483,7 @@ class PeerConnectionV2 extends StateMachine {
     const { iceConnectionState } = this._peerConnection;
 
-    this._log.debug(`ICE connection state is "${iceConnectionState}"`);
+    this._log.debug(`ICE connection state is "${this._lastIceConnectionState}" => "${iceConnectionState}"`);
```
@makarandp0 I think we should not change the log statement here since `_lastIceConnectionState` can be null.

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,100 @@
+  _getIceConnectionStats() {
+    return this._peerConnection.getStats().then(stats => Array.from(stats.values()).find(stat => {
+      return stat.type === 'candidate-pair' && stat.nominated;
+    })).then((activeCandidatePairStats) => {
```
@makarandp0 Nitpick: I think we can return the whole RTCIceCandidatePairStats here and get rid of the `then()` callback. Also, I think you can get rid of the braces around `activeIceCandidatePairStats`.

makarandp0 replied:
imo having a limited fields returned from _getIceConnectionStats, helps with unit test/mocking of this function. I will  update  to get rid of braces.

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  /**
+   * Stop monitoring the ICE connection state
```
@makarandp0 Nitpick: Add a `.` at the end?

manjeshbhargav commented on lib/signaling/v2/peerconnection.js:
```diff
@@ -290,7 +299,8 @@ class PeerConnectionV2 extends StateMachine {
    */
   get iceConnectionState() {
-    return this._peerConnection.iceConnectionState;
+    return (this._isIceConnectionInactive && this._peerConnection.iceConnectionState === 'disconnected')
```
@makarandp0 Nitpick: I don't think we need the `()` around the condition. Also, can you indent the `?` with just two spaces after `return` like this:
```js
return this._isIceConnectionInactive && this._peerConnection.iceConnectionState === 'disconnected'
  ? 'failed' : this._peerConnection.iceConnectionState;
```
?

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -46,13 +46,7 @@ class IceConnectionMonitor {
-        bytesSent: activeCandidatePairStats.bytesSent
-      };
-    }).catch(() => {
+    })).catch(() => {
```
@makarandp0 I'll leave it up to you, but the catch callback can also be the second argument for `then()`:
```js
.then(onResolved, onRejected);
```

makarandp0 replied:
@manjeshbhargav, I do see usage of 2nd param of `.then` in our codebase, and thinking we should move to using .catch instead - 1st type of usage could be problematic. More about this antipattern at https://www.datchley.name/promise-patterns-anti-patterns/

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,94 @@
+
+  /**
+   * Get ICE connection stats, and extract received and send bytes.
+   * @returns Promise<?{timestamp, bytesReceived, bytesSent}>
```
@makarandp0 `@returns Promise<?RTCIceCandidatePairStats>`?

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,94 @@
+          return;
+        }
+
+        // if change detected in bytesReceived mark it as lastActiveiy.
```
@makarandp0 I think the code is self-explanatory, so we don't need the comment here.

manjeshbhargav commented on lib/signaling/v2/iceconnectionmonitor.js:
```diff
@@ -0,0 +1,94 @@
+          this._lastActivity = iceStats;
+        }
+
+        // after inactivity threshold fire inactiveCallback.
```
@makarandp0 I think the code is self-explanatory, so we don't need the comment here.
